File: post-install/app/js/wallpaper.js
// Preia fundalul curent setat în KDE Plasma al utilizatorului și îl aplică
// ca fundal al aplicației (în loc de imaginea statică din resources/),
// prin variabila CSS --wallpaper (vezi assistant-styles.css).
(function () {
  var fs = require('fs');
  var path = require('path');
  var os = require('os');

  // Extrage Image= dintr-o secțiune INI ale cărei chei de secțiune se termină
  // exact cu `sectionSuffix` (ex: "Wallpaper][org.kde.image][General" pentru
  // formatul din plasma-org.kde.plasma.desktop-appletsrc, sau "Wallpaper"
  // pentru formatul flat din pachetele Look-and-Feel).
  function readImageFromIniFile(filePath, sectionSuffix) {
    var content;
    try {
      content = fs.readFileSync(filePath, 'utf8');
    } catch (e) {
      return null;
    }

    var lines = content.split('\n');
    var inSection = false;
    var found = null;
    for (var i = 0; i < lines.length; i++) {
      var line = lines[i];
      var sectionMatch = line.match(/^\[(.+)\]$/);
      if (sectionMatch) {
        inSection = sectionMatch[1] === sectionSuffix || sectionMatch[1].indexOf(']' + sectionSuffix) !== -1 || sectionMatch[1].slice(-sectionSuffix.length) === sectionSuffix;
        continue;
      }
      if (inSection) {
        var m = line.match(/^Image=(.*)$/);
        // Prima valoare găsită e suficientă pentru un fundal reprezentativ.
        if (m && !found) found = m[1].trim();
      }
    }
    return found;
  }

  // Config-ul per-utilizator, dacă a schimbat vreodată wallpaper-ul explicit.
  function readUserWallpaperOverride() {
    var configPath = path.join(os.homedir(), '.config', 'plasma-org.kde.plasma.desktop-appletsrc');
    return readImageFromIniFile(configPath, 'Wallpaper][org.kde.image][General');
  }

  // Dacă nu există override, wallpaper-ul vine din tema Look-and-Feel (Global
  // Theme) activă — nescris niciodată explicit în config-ul de mai sus.
  function readLookAndFeelWallpaper() {
    var kdeglobals = path.join(os.homedir(), '.config', 'kdeglobals');
    var content;
    try {
      content = fs.readFileSync(kdeglobals, 'utf8');
    } catch (e) {
      return null;
    }
    var m = content.match(/^LookAndFeelPackage=(.+)$/m);
    if (!m) return null;
    var themeId = m[1].trim();

    var defaultsPaths = [
      path.join(os.homedir(), '.local/share/plasma/look-and-feel', themeId, 'contents/defaults'),
      path.join('/usr/share/plasma/look-and-feel', themeId, 'contents/defaults')
    ];
    for (var i = 0; i < defaultsPaths.length; i++) {
      var val = readImageFromIniFile(defaultsPaths[i], 'Wallpaper');
      if (val) return val;
    }
    return null;
  }

  function fileUriToPath(uri) {
    if (uri.indexOf('file://') === 0) {
      try {
        return decodeURIComponent(uri.slice('file://'.length));
      } catch (e) {
        return uri.slice('file://'.length);
      }
    }
    return uri;
  }

  // O valoare Image= poate fi: un file:// URI, o cale absolută, sau doar
  // id-ul pachetului de wallpaper (ex: "Next") căutat în directoarele
  // standard de wallpapers.
  function candidatePaths(value) {
    if (value.indexOf('file://') === 0 || value.indexOf('/') === 0) {
      return [fileUriToPath(value)];
    }
    return [
      path.join(os.homedir(), '.local/share/wallpapers', value),
      path.join('/usr/share/wallpapers', value)
    ];
  }

  // Pachetele de wallpaper KDE sunt directoare cu mai multe rezoluții în
  // contents/images/; alegem cea mai mare disponibilă.
  function resolveImageFile(p) {
    var stat;
    try {
      stat = fs.statSync(p);
    } catch (e) {
      return null;
    }
    if (stat.isFile()) return p;
    if (!stat.isDirectory()) return null;

    var imagesDir = path.join(p, 'contents', 'images');
    var candidates;
    try {
      candidates = fs.readdirSync(imagesDir);
    } catch (e) {
      return null;
    }
    if (candidates.length === 0) return null;

    candidates.sort(function (a, b) {
      function area(name) {
        var m = name.match(/(\d+)x(\d+)/);
        return m ? parseInt(m[1], 10) * parseInt(m[2], 10) : 0;
      }
      return area(b) - area(a);
    });
    return path.join(imagesDir, candidates[0]);
  }

  function resolveWallpaper(value) {
    var candidates = candidatePaths(value);
    for (var i = 0; i < candidates.length; i++) {
      var resolved = resolveImageFile(candidates[i]);
      if (resolved) return resolved;
    }
    return null;
  }

  function toFileUrl(p) {
    return 'file://' + p.split(path.sep).map(encodeURIComponent).join('/');
  }

  var raw = readUserWallpaperOverride() || readLookAndFeelWallpaper();
  if (!raw) return;

  var resolved = resolveWallpaper(raw);
  if (!resolved) return;

  document.documentElement.style.setProperty('--wallpaper', 'url("' + toFileUrl(resolved) + '")');
})();

File: post-install/gtkapp/wallpaper.py
"""Port of app/js/wallpaper.js: find the user's current KDE Plasma wallpaper,
falling back to the bundled resources/background image."""
import os
import re

_APP_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DEFAULT_WALLPAPER = os.path.join(_APP_DIR, "app", "resources", "background.png")

_SECTION_RE = re.compile(r"^\[(.+)\]$")
_IMAGE_RE = re.compile(r"^Image=(.*)$")


def _read_image_from_ini_file(file_path, section_suffix):
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
    except OSError:
        return None

    in_section = False
    found = None
    for line in content.split("\n"):
        m = _SECTION_RE.match(line)
        if m:
            name = m.group(1)
            in_section = (
                name == section_suffix
                or (("]" + section_suffix) in name)
                or name[-len(section_suffix):] == section_suffix
            )
            continue
        if in_section:
            m2 = _IMAGE_RE.match(line)
            if m2 and not found:
                found = m2.group(1).strip()
    return found


def _read_user_wallpaper_override():
    config_path = os.path.join(
        os.path.expanduser("~"), ".config", "plasma-org.kde.plasma.desktop-appletsrc"
    )
    return _read_image_from_ini_file(config_path, "Wallpaper][org.kde.image][General")


def _read_look_and_feel_wallpaper():
    kdeglobals = os.path.join(os.path.expanduser("~"), ".config", "kdeglobals")
    try:
        with open(kdeglobals, "r", encoding="utf-8") as f:
            content = f.read()
    except OSError:
        return None
    m = re.search(r"^LookAndFeelPackage=(.+)$", content, re.MULTILINE)
    if not m:
        return None
    theme_id = m.group(1).strip()

    defaults_paths = [
        os.path.join(
            os.path.expanduser("~"),
            ".local/share/plasma/look-and-feel",
            theme_id,
            "contents/defaults",
        ),
        os.path.join("/usr/share/plasma/look-and-feel", theme_id, "contents/defaults"),
    ]
    for p in defaults_paths:
        val = _read_image_from_ini_file(p, "Wallpaper")
        if val:
            return val
    return None


def _file_uri_to_path(uri):
    if uri.startswith("file://"):
        from urllib.parse import unquote
        return unquote(uri[len("file://"):])
    return uri


def _candidate_paths(value):
    if value.startswith("file://") or value.startswith("/"):
        return [_file_uri_to_path(value)]
    return [
        os.path.join(os.path.expanduser("~"), ".local/share/wallpapers", value),
        os.path.join("/usr/share/wallpapers", value),
    ]


def _resolve_image_file(p):
    if os.path.isfile(p):
        return p
    if not os.path.isdir(p):
        return None

    images_dir = os.path.join(p, "contents", "images")
    try:
        candidates = os.listdir(images_dir)
    except OSError:
        return None
    if not candidates:
        return None

    def area(name):
        m = re.search(r"(\d+)x(\d+)", name)
        return int(m.group(1)) * int(m.group(2)) if m else 0

    candidates.sort(key=area, reverse=True)
    return os.path.join(images_dir, candidates[0])


def _resolve_wallpaper(value):
    for candidate in _candidate_paths(value):
        resolved = _resolve_image_file(candidate)
        if resolved:
            return resolved
    return None


def find_wallpaper():
    """Return an absolute path to the live wallpaper, or the bundled default."""
    raw = _read_user_wallpaper_override() or _read_look_and_feel_wallpaper()
    if raw:
        resolved = _resolve_wallpaper(raw)
        if resolved:
            return resolved
    return DEFAULT_WALLPAPER
